lib/services/statistics_service.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require "json"
+require "date"
+
+class StatisticsService
+  def initialize(data_file_path)
+    @data_file_path = data_file_path
+  end
+
+  # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+  def work
+    users = []
+    sessions = []
+    browsers = []
+
+    File.readlines(data_file_path, chomp: true).each do |line|
+      cols = line.split(",")
+
+      next users << parse_user(cols) if cols[0] == "user"
+
+      session = parse_session(cols)
+      browsers << session["browser"]
+      sessions << session
+    end
+
+    grouped_sessions = sessions.group_by { |session| session["user_id"] }
+    unique_browsers = browsers.uniq.sort
```

Comment:
Можно сразу собирать в Set / SortedSet

Reply:
I try to use a set for saving unique_browsers but it is not a better solution for me. I see it in the results of speeds. I didn't see source files but my guess is each time when I use `set << some_valie` it is `array << some_value + array.uniq`. Because each time when we try to append some value, we should check if it is a unique value or not = O(n). Bot based on this [response](https://stackoverflow.com/questions/14200540/what-is-the-time-complexity-of-arrayuniq-method-in-ruby) and see the same problem for some [people](https://www.reddit.com/r/ruby/comments/n09a56/set_vs_arrayuniq/). I can say unique for an array is better than using a set where we append data each time. `n*Array.<<` + `Array.uniq`  = O(n) is better `n*Set.<<`